### PR TITLE
Fix Issue #181

### PR DESCRIPTION
**Pull request by Davknapp**
Restructured the basic-hcube example. 
This is currently a draft, if the quality meets our standards I suggest to delete the other basic-examples. Then, I will convert this into an ordinary PR.

**_All these boxes must be checked by the reviewers before merging the pull request:_**

- [x] The author added a BSD statement to `doc/` (or already has one)
- [x] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)

  If the Pull request indroduces code that is not covered by the github action (for example coupling with a new library):
  - [x] Should this use case be added to the github action?
  - [x] If not, does the specific use case compile and all tests pass (check manually)

- [x] All tests pass (in various configurations, this should be executed automatically in a github action)
- [x] New source/header files are properly added to the Makefiles
- [x] The reviewer executed the new code features at least once and checked the results manually
- [x] The code is covered in an existing or new test case
- [x] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [x] The code is well documented
- [x] All function declarations, structs/classes and their members have a proper doxygen documentation
- [x] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)
- [x] Testing of this template: If you feel something is missing from this list, contact the developers

### CODE REVIEW

File: example/basic/basic_hcube.cxx
Comment (holke) [inline example/basic/basic_hcube.cxx:30]:
```suggestion
 * \param[in] dim           The dimension of the example. 1 <= \a dim <= 3
```

File: example/basic/basic_hcube.cxx
Comment (holke) [inline example/basic/basic_hcube.cxx:31]: Please specify the concrete behavior of the parameter.
"If true (non-zero) the cmesh will be partitioned across the processes..."

File: example/basic/basic_hcube.cxx
Comment (holke) [inline example/basic/basic_hcube.cxx:32]:
```suggestion
static t8_cmesh_t
```
Functions that are not declared in a header/not used in another file should be static

File: example/basic/basic_hcube.cxx
Comment (holke) [inline example/basic/basic_hcube.cxx:29]: What kind of cmesh is constructed?
How do the parameters affect how the cmesh looks like?

File: example/basic/basic_hcube.cxx
Comment (holke) [inline example/basic/basic_hcube.cxx:31]: \return is missing from the doc

File: example/basic/basic_hcube.cxx
Comment (holke) [inline example/basic/basic_hcube.cxx:66]:
```suggestion
 * x-coordinate wrt to its level is refined up to the maximal refinement level */
```

Comment (holke) [inline example/basic/basic_hcube.cxx:66]: Coordinates are double valued. What does uneven mean?

**Review by holke — COMMENTED** [review]
Just a few quick first comments, since I had 5 minutes to spare.
Complete review will follow

File: example/basic/basic_hcube.cxx
Comment (Davknapp) [reply example/basic/basic_hcube.cxx]: I meant the coordinate in the reference element. The reference coordinate is shifted by 
`maximum_level - level_of_the_element`
and if that coordinate is odd the element gets refined. 
I updated that comment.

File: example/basic/basic_hcube.cxx
Comment (holke) [inline example/basic/basic_hcube.cxx:191]:
```suggestion
              "cubical mesh, which adaptively refines up to level 5.\n"
```

File: example/basic/basic_hcube.cxx
Comment (holke) [inline example/basic/basic_hcube.cxx:190]: why "up to"?
I do not really understand the phrasing here.

File: example/basic/basic_hcube.cxx
Comment (holke) [inline example/basic/basic_hcube.cxx:214]: If the user does not have a choice for the maximum refinement level (which is okay, since it remove complexity), then we do not need the maximum refinement level as user data of the forest and should remove it.
Instead we can just set the maximum level variable in the refine callback to 5.
Thus, we remove again a piece of complexity.

File: example/basic/basic_hcube.cxx
Comment (holke) [inline example/basic/basic_hcube.cxx:212]: Add the possible options in the description. 1 <= dim <= 3

File: example/basic/basic_hcube.cxx
Comment (holke) [inline example/basic/basic_hcube.cxx:214]:
```suggestion
                         "Additionally 2:1 balance the forest.");
```

**Review by holke — CHANGES_REQUESTED** [review]
Looks good. Thanks for working on this so quickly.

I have some request how we can improve this example.

Imho we should keep the complexity as low as possible.

File: example/basic/basic_hcube.cxx
Comment (Davknapp) [reply example/basic/basic_hcube.cxx:91]: Sure, I'll add it

File: example/basic/basic_hcube.cxx
Comment (Davknapp) [reply example/basic/basic_hcube.cxx:139]: I think we could wrap an output check in a small seperate function. That way we do not duplicate code and check everytime if the output was generated.

Comment (Davknapp) [reply example/basic/basic_hcube.cxx:139]: Just figured out that t8_forest_write_vtk is a void function - hence we can not check for a success.

**Comment by holke** [comment]
I vote for calling this new example "t8_basic" now that the old is deleted.

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:65]:
```suggestion
/* The adapt Callback used in this example. Every element that has an odd
```

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:66]: this comment needs to be updated

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:79]:
```suggestion
    /* We do not refine if the maximum level is reached */
```

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:82]: We should use `t8_forest_element_centroid` here.

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:97]:
```suggestion
/** This function creates, adaptively refines, partitions and optionally balances a
```

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:101]: The comment regarding do_balance is wrong

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:113]:
```suggestion
  /* Create the cmesh */
```

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:124]:
```suggestion
    t8_errorf ("Error when trying to write cmesh to %s.\n",
```

We need to be careful in such situations. 
1) in case of an error, an error logging function should be called
2) the global logging functions only print on the main process. But the writing operation could be succesful on the main process and only fail on another proc. In that case, we would not see the error output when using a global output function.

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:141]:
```suggestion
  snprintf (vtuname, BUFSIZ, "forest_basic_%id_uniform", dim);
```

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:159]:
```suggestion
  snprintf (vtuname, BUFSIZ, "forest_hypercube_%id_adapt", dim);
```

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:193]:
```suggestion
              "cubical mesh, which is then adaptively refined with "
```

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:196]:
```suggestion
              "and whether it should be 2:1 balanced or not.\n\n%s\n", usage);
```

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:209]: Initialization should happen before calling t8_debugf

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:215]:
```suggestion
                      "The dimension of the mesh. Choose 1 <= dimension <= 3.");
```

**Review by holke — CHANGES_REQUESTED** [review]
Looks good now. Only minor changes.
I need to run it myself once before i merge.

File: example/basic/basic_hcube.cxx
Comment (holke) [reply example/basic/basic_hcube.cxx]: See #190

**Review by holke — COMMENTED** [review]
looks good.
will merge when i ran it myself once.

Thanks :)

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:121]:
```suggestion
  t8_global_productionf ("Constructing %i dimensional hypercube mesh \n", dim);
```

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:233]:
```suggestion
```

File: example/basic/t8_basic.cxx
Comment (holke) [inline example/basic/t8_basic.cxx:229]:
```suggestion
    sc_options_parse (t8_get_package_id (), SC_LP_ERROR, opt, argc, argv);
    
    /* Print usage */
    sc_options_print_usage (t8_get_package_id (), SC_LP_ERROR, opt, NULL);
```

**Review by holke — COMMENTED** [review]
minor edits

**Pull request merged**
